From be919cad35548f616ccce979b4ab3d4ed6e1e28d Mon Sep 17 00:00:00 2001
From: David Disseldorp <ddiss@suse.de>
Date: Sat, 22 Aug 2015 18:43:42 +0200
Subject: [PATCH] rbd: add rbd_dev_getxattr() helper
Patch-mainline: Not yet, SES2 clustered LIO/RBD
References: fate#318836

Signed-off-by: David Disseldorp <ddiss@suse.de>
[jeffm: folded in patches.suse/rbd-do-away-with-obj_request-in-getxattr.patch
 for sorted series]
---
 drivers/block/rbd.c         |   68 ++++++++++++++++++++++++++++++++++++++++++++
 include/linux/ceph/librbd.h |    2 +
 2 files changed, 70 insertions(+)

--- a/drivers/block/rbd.c
+++ b/drivers/block/rbd.c
@@ -1846,6 +1846,7 @@ static void rbd_osd_req_callback(struct
 		break;
 	case CEPH_OSD_OP_SETXATTR:
 	case CEPH_OSD_OP_CMPXATTR:
+	case CEPH_OSD_OP_GETXATTR:
 		obj_request_done_set(obj_request);
 		break;
 	default:
@@ -4018,6 +4019,73 @@ out:
 }
 EXPORT_SYMBOL(rbd_dev_cmpsetxattr);
 
+int rbd_dev_getxattr(struct rbd_device *rbd_dev, char *key, int max_val_len,
+		     void **_val, int *val_len)
+{
+	struct ceph_osd_client *osdc = &rbd_dev->rbd_client->client->osdc;
+	struct ceph_osd_request *req;
+	int page_count;
+	struct page **pages = NULL;
+	void *val;
+	int ret;
+
+	BUG_ON(!key);
+	BUG_ON(!_val || !val_len);
+
+	req = ceph_osdc_alloc_request(osdc, NULL, 1, false, GFP_KERNEL);
+	if (!req)
+		return -ENOMEM;
+
+	ceph_oid_copy(&req->r_base_oid, &rbd_dev->header_oid);
+	ceph_oloc_copy(&req->r_base_oloc, &rbd_dev->header_oloc);
+	req->r_flags = CEPH_OSD_FLAG_READ;
+
+	ret = ceph_osdc_alloc_messages(req, GFP_KERNEL);
+	if (ret)
+		goto out_req;
+
+	ret = osd_req_op_xattr_init(req, 0,
+				    CEPH_OSD_OP_GETXATTR,
+				    key, NULL, 0, 0, 0);
+	if (ret)
+		goto out_req;
+
+	page_count = calc_pages_for(0, max_val_len);
+	pages = ceph_alloc_page_vector(page_count, GFP_KERNEL);
+	if (IS_ERR(pages)) {
+		ret = PTR_ERR(pages);
+		goto out_req;
+	}
+
+	osd_req_op_xattr_response_data_pages(req, 0,
+					     pages, max_val_len,
+					     0, false, false);
+
+	ceph_osdc_start_request(osdc, req, false);
+	ret = ceph_osdc_wait_request(osdc, req);
+	if (ret < 0)
+		goto out_req;
+
+	rbd_assert(ret <= (u64)max_val_len);
+
+	val = kmalloc(ret, GFP_KERNEL);
+	if (!val) {
+		ret = -ENOMEM;
+		goto out_req;
+	}
+
+	ceph_copy_from_page_vector(pages, val, 0, ret);
+	*_val = val;
+	*val_len = ret;
+
+	ret = 0;
+
+out_req:
+	ceph_osdc_put_request(req);
+	return ret;
+}
+EXPORT_SYMBOL(rbd_dev_getxattr);
+
 /*
  * TODO: remove me or move to debugfs for final merge.
  */
--- a/include/linux/ceph/librbd.h
+++ b/include/linux/ceph/librbd.h
@@ -177,5 +177,7 @@ extern int rbd_dev_setxattr(struct rbd_d
 extern int rbd_dev_cmpsetxattr(struct rbd_device *rbd_dev, char *key,
 			       void *oldval, int oldval_len, void *newval,
 			       int newval_len);
+extern int rbd_dev_getxattr(struct rbd_device *rbd_dev, char *key, int max_val_len,
+			    void **_val, int *val_len);
 
 #endif
